sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -848,25 +848,40 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
 
 ssize_t coap_opt_add_uquery(coap_pkt_t *pdu, const char *key, const char *val)
 {
+    return coap_opt_add_uquery2(pdu, key, strlen(key), val, val ? strlen(val) : 0);
+}
+
+ssize_t coap_opt_add_uquery2(coap_pkt_t *pkt, const char *key, size_t key_len,
+                             const char *val, size_t val_len)
+{
+    assert(pkt);
+    assert(key);
+    assert(key_len);
+
     char qs[NANOCOAP_QS_MAX];
-    size_t len = strlen(key);
-    size_t val_len = (val) ? (strlen(val) + 1) : 0;
 
-    /* test if the query string fits, account for the zero termination */
-    if ((len + val_len + 1) >= NANOCOAP_QS_MAX) {
+    /* account for '=' */
+    if (val) {
+        val_len = val_len + 1;
+    }
+    else {
+        val_len = 0;
+    }
+
+    /* test if the query string fits, account for the zero termination
+     * Account for '\0'*/
+    if ((key_len + val_len + 1) >= NANOCOAP_QS_MAX) {
         return -1;
     }
 
-    memcpy(&qs[0], key, len);
-    if (val) {
-        qs[len] = '=';
-        /* the `=` character was already counted in `val_len`, so subtract it here */
-        memcpy(&qs[len + 1], val, (val_len - 1));
-        len += val_len;
+    memcpy(&qs[0], key, key_len);
+    if (val && val_len > 0) {
```

Comment:
Either just use `(val && val_len)` or add the extra parens `(val && (val_len > 0))`

Reply:
changed to `(val && val_len)`.